### PR TITLE
Adding memory macros for TSMC 28

### PR DESCRIPTION
**Pull request by dpetrisko**
Adds memory macros for TSMC28. Part of a decomposition of #492

### CODE REVIEW

File: hard/tsmc_28/bsg_mem/bsg_mem_1r1w_sync_macros.vh [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+             ,.QA       (               )                       \
+                                                                \
+             ,.AB       ( r_addr_i      )                       \
+             ,.DB       ( '1            )                       \
+             ,.WEBB     ( '1            )                       \
+             ,.CEBB     ( ~r_v_i        )                       \
+             ,.CLKB     ( clk_i         )                       \
+             ,.QB       ( r_data_o      )                       \
+                                                                \
+             ,.WTSEL    ( 2'b01         )                       \
```
Comment (taylor-bsg) [inline hard/tsmc_28/bsg_mem/bsg_mem_1r1w_sync_macros.vh:48]: Add comment justifying selection of these chickenbits

File: hard/tsmc_28/bsg_mem/bsg_mem_1r1w_sync_macros.vh [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+             ,.CLKA     ( clk_i         )                       \
+             ,.QA       (               )                       \
+                                                                \
+             ,.AB       ( r_addr_i      )                       \
+             ,.DB       ( '1            )                       \
+             ,.WEBB     ( '1            )                       \
+             ,.CEBB     ( ~r_v_i        )                       \
+             ,.CLKB     ( clk_i         )                       \
+             ,.QB       ( r_data_o      )                       \
+                                                                \
```
Comment (taylor-bsg) [inline hard/tsmc_28/bsg_mem/bsg_mem_1r1w_sync_macros.vh:73]: Justify chickenbits

File: hard/tsmc_28/bsg_mem/bsg_mem_1r1w_sync_mask_write_bit_macros.vh [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+                                                                \
+             ,.AB       ( r_addr_i      )                       \
+             ,.DB       ( '1            )                       \
+             ,.BWEBB    ( '1            )                       \
+             ,.WEBB     ( '1            )                       \
+             ,.CEBB     ( ~r_v_i        )                       \
+             ,.CLKB     ( clk_i         )                       \
+             ,.QB       ( r_data_o      )                       \
+                                                                \
+             ,.WTSEL    ( 2'b01         )                       \
```
Comment (taylor-bsg) [inline hard/tsmc_28/bsg_mem/bsg_mem_1r1w_sync_mask_write_bit_macros.vh:43]: Justify chickenbits

File: hard/tsmc_28/bsg_mem/bsg_mem_1r1w_sync_mask_write_bit_macros.vh [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+             ,.QA       (               )                       \
+                                                                \
+             ,.AB       ( r_addr_i      )                       \
+             ,.DB       ( '1            )                       \
+             ,.BWEBB    ( '1            )                       \
+             ,.WEBB     ( '1            )                       \
+             ,.CEBB     ( ~r_v_i        )                       \
+             ,.CLKB     ( clk_i         )                       \
+             ,.QB       ( r_data_o      )                       \
+                                                                \
```
Comment (taylor-bsg) [inline hard/tsmc_28/bsg_mem/bsg_mem_1r1w_sync_mask_write_bit_macros.vh:70]: Justify chickenbits

File: hard/tsmc_28/bsg_mem/bsg_mem_1rw_sync_macros.vh [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    begin: macro                                     \
+          tsmc28_1rw_d``words``_w``bits``_``tag``_1hdsram mem  \
+            (                                                   \
+              .CLK      ( clk_i         )                       \
+             ,.CEB      ( ~v_i          )                       \
+             ,.WEB      ( ~w_i          )                       \
+             ,.A        ( addr_i        )                       \
+             ,.D        ( data_i        )                       \
+             ,.Q        ( data_o        )                       \
+             ,.RTSEL    ( 2'b01         )                       \
```
Comment (taylor-bsg) [inline hard/tsmc_28/bsg_mem/bsg_mem_1rw_sync_macros.vh:43]: Chick

File: hard/tsmc_28/bsg_mem/bsg_mem_1rw_sync_mask_write_bit_macros.vh [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+          tsmc28_1rw_d``words``_w``bits``_``tag``_bit_1hdsram mem  \
+            (                                                   \
+              .CLK      ( clk_i         )                       \
+             ,.CEB      ( ~v_i          )                       \
+             ,.WEB      ( ~w_i          )                       \
+             ,.BWEB     ( ~w_mask_i     )                       \
+             ,.A        ( addr_i        )                       \
+             ,.D        ( data_i        )                       \
+             ,.Q        ( data_o        )                       \
+             ,.RTSEL    ( 2'b01         )                       \
```
Comment (taylor-bsg) [inline hard/tsmc_28/bsg_mem/bsg_mem_1rw_sync_mask_write_bit_macros.vh:48]: Chick

**Review by taylor-bsg — CHANGES_REQUESTED** [review]

**Comment by dpetrisko** [comment]
Added comments for all chicken bits

**Comment by dpetrisko** [comment]
Good to go then?

File: hard/tsmc_28/bsg_mem/bsg_mem_3r1w_sync_macros.vh [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+              .AA       ( w_addr_i      )                       \
+             ,.DA       ( w_data_i      )                       \
+             ,.WEBA     ( ~w_v_i        )                       \
+             ,.CEBA     ( ~w_v_i        )                       \
+             ,.CLK      ( clk_i         )                       \
+             ,.QA       (               )                       \
+                                                                \
+             ,.AB       ( r0_addr_i     )                       \
+             ,.DB       ( '1            )                       \
+             ,.WEBB     ( '1            )                       \
```
Comment (tommydcjung) [inline hard/tsmc_28/bsg_mem/bsg_mem_3r1w_sync_macros.vh:131]: how many bits is this supposed to be? maybe be more explicit about the width?

Comment (dpetrisko) [reply hard/tsmc_28/bsg_mem/bsg_mem_3r1w_sync_macros.vh]: made explicit!

**Comment by dpetrisko** [comment]
good to go?

**Review by tommydcjung — APPROVED** [review]

**Comment by dpetrisko** [comment]
Thanks!

**Pull request merged**